src/colaa.cpp

```diff
@@ -0,0 +1,420 @@
+    // that's non-POSIX (doesn't work on OS X, for example).
+    struct timeval tv;
+    tv.tv_sec = 0;
+    tv.tv_usec = 100000;
+
+    logDebug("entering select()", tv.tv_usec);
+    int retval = select(socket_fd_ + 1, &rfds, NULL, NULL, &tv);
+    logDebug("returned %d from select()", retval);
+    if (retval)
+    {
+      buffer_->readFrom(socket_fd_);
+
+      // Will return pointer if a complete message exists in the buffer,
+      // otherwise will return null.
+      char* buffer_data = buffer_->getNextBuffer();
+
+      if (buffer_data)
+      {
+        parse_scan_data(buffer_data, scan_data);
+        buffer_->popLastBuffer();
+        return true;
+      }
+    }
+    else
+    {
+      // Select timed out or there was an fd error.
+      return false;
+    }
+  }
+}
+
+void CoLaA::do_login(std::string user_class, std::string password)
+{
+   std::string command = LOGIN_COMMAND + " " + user_class + " " + password;
+   send_command(command);
+}
+
+ScanConfig CoLaA::parse_scan_cfg(char *buf, size_t len)
+{
+  ScanConfig cfg;
+  next_token(&buf); // Command type
+  next_token(&buf); // Command name
+  next_token(&buf, cfg.scan_frequency);
+  next_token(&buf, cfg.num_sectors);
+  next_token(&buf, cfg.angualar_resolution);
+  next_token(&buf, cfg.start_angle);
+  next_token(&buf, cfg.stop_angle);
+  return cfg;
+}
+
+std::string CoLaA::build_scan_cfg(const ScanConfig &cfg) const
+{
+  if (cfg.num_sectors > 1)
+  {
+    logWarn("This method does not support configuring multiple sectors");
+  }
+  std::stringstream ss;
+  ss << std::uppercase << std::hex << cfg.scan_frequency << " +" << std::dec << std::min(cfg.num_sectors, (int16_t)1) << " "
+     << std::hex << cfg.angualar_resolution << " " << cfg.start_angle << " " << cfg.stop_angle;
+  logDebug("TX: %s", ss.str().c_str());
+  return ss.str();
+}
+
+std::string CoLaA::build_scan_data_cfg(const ScanDataConfig &cfg) const
+{
+  std::stringstream ss;
+  ss << build_scan_data_cfg_output_channel(cfg.output_channel);
+  ss << " " << cfg.remission;
+  ss << " " << cfg.resolution;
+  ss << " 0"; // Resolution, always 0
+  ss << " " << build_scan_data_cfg_encoder(cfg.encoder);
+  ss << " " << cfg.position;
+  ss << " " << cfg.device_name;
+  ss << " " << cfg.comment;
+  ss << " " << cfg.timestamp;
+  ss << " +" << cfg.output_interval;
+  return ss.str();
+}
+
+std::string CoLaA::build_scan_data_cfg_output_channel(int ch) const
+{
+  std::stringstream ss;
+  ss << std::setw(2) << std::setfill('0') << ch << " 00";
+  return ss.str();
+}
+
+std::string CoLaA::build_scan_data_cfg_encoder(int enc) const
+{
+  if (enc)
+  {
+    std::stringstream ss;
+    ss << std::setw(2) << std::setfill('0') << enc << " 00";
+    return ss.str();
+  }
+  return "00 00"; // Data sheet says "No encoder: 0, but that produces an error"
+}
+
+// Dangerous hack!
+void copy_vector(std::vector<uint8_t> &src, uint16_t *dest)
+{
+  for (size_t i = 0; i < src.size(); ++i)
+  {
+    dest[i] = src[i];
+  }
+}
```

Comment:
It's no longer used, was used to copy stuff into the intermediate scanData structure.